[Qt] allow banning and unbanning over UI->peers table

jonasschnelli opened the pull request:
This adds 4 ban options to the peers tables context menu (1h, 24h, 7d, 365d). If there are active bans, a table with banned subnets will be shown.
Sync between the rpc bans and the ui bans are guaranteed over a new `uiInterface` signal.

Screenshots:
![bildschirmfoto-2015-06-20-um-21 58 44](https://cloud.githubusercontent.com/assets/178464/8269026/ac00e74e-1797-11e5-8018-fa8b447292de.png)
![bildschirmfoto-2015-06-20-um-21 58 59](https://cloud.githubusercontent.com/assets/178464/8269027/ac287b24-1797-11e5-92b3-9d6c17d00fde.png)

jonasschnelli committed:
[Qt] add ban functions to peers window

add ban option for peer context menu (1h, 24h, 7d, 1y).

jonasschnelli committed:
[Qt] add banlist table below peers table

jonasschnelli committed:
[Qt] add ui signal for banlist changes

jonasschnelli committed:
[Qt] add context menu with unban option to ban table

jonasschnelli committed:
[Qt] banlist, UI optimizing and better signal handling

jonasschnelli commented on src/rpcnet.cpp:
```diff
@@ -527,6 +528,7 @@ UniValue setban(const UniValue& params, bool fHelp)
             throw JSONRPCError(RPC_MISC_ERROR, "Error: Unban failed");
     }
 
+    uiInterface.BannedListChanged();
```
I'm not sure if the locking of a signal emit is done within boost::signal of if this needs explicit locking of `uiInterface`?

Diapolo replied:
Perhaps @laanwj can comment on this?

jonasschnelli commented on src/qt/bantablemodel.cpp:
```diff
@@ -0,0 +1,192 @@
+        case Bantime:
+            //show time in users local timezone, not 64bit compatible!
+            //TODO find a way to support 64bit timestamps
+            boost::posix_time::ptime pt1 = boost::posix_time::from_time_t(rec->bantil);
```
i recognized that `boost::posix_time::from_time_t` is not 64bit capable. Maybe someone has a idea how to fix this.

Diapolo replied:
What about ``QDateTime::fromTime_t``?

jonasschnelli replied:
I think this won't work either. The [docs](http://doc.qt.io/qt-5.4/qdatetime.html#fromTime_t) are saying: `QDateTime QDateTime::fromTime_t(uint seconds)` (uint!).

jonasschnelli replied:
Found a fix! Pushed.

Diapolo commented on src/qt/bantablemodel.cpp:
```diff
@@ -0,0 +1,192 @@
+    /** Pull a full list of banned nodes from CNode into our cache */
+    void refreshBanlist()
+    {
+        std::map<CSubNet, int64_t> banMap;
```
I was thinking about a typedef for this, what do you mean?

jonasschnelli replied:
`typedef`s are useful but can also make reading the code more complicate. In this case (simple map) i don't see a usecase for a typedef.

prestotron replied:
Why would it make it more complicated? Worse comes to worse, you comment it, which is not really needed anyways.

Diapolo commented on src/qt/bantablemodel.cpp:
```diff
@@ -0,0 +1,192 @@
+        cachedBanlist.reserve(banMap.size());
+#endif
+        std::map<CSubNet, int64_t>::iterator iter;
+        for (iter = banMap.begin(); iter != banMap.end(); ++iter) {
```
Could this be a Qt foreach? Would also be easy to convert to C++11 then :)?

Diapolo commented on src/qt/bantablemodel.cpp:
```diff
@@ -0,0 +1,192 @@
+        }
+    }
+
+    int size()
```
Could this be ``int size() const``?

Diapolo commented on src/qt/bantablemodel.cpp:
```diff
@@ -0,0 +1,192 @@
+    {
+        if(idx >= 0 && idx < cachedBanlist.size()) {
+            return &cachedBanlist[idx];
+        } else {
```
That else can be removed. I see you copied most of the code here from peertablemodel.cpp where all these suggestions would also fit and perhaps we should clean it up, to keep the codebase close together?

jonasschnelli replied:
Not sure if this else can be removed. What if the `QTableView` requests a row at index which is not in the cache `QList`. Obviously it should not happen because the table size is driven by the size of the cache `QList`. But I prefer keeping this.

Diapolo replied:
Not sure if you're right here, when the if clause isn't true we always run into the else (as we have a return in the if we won't reach the return 0, if the clause is false). I'd rather remove it...

Diapolo commented on src/qt/bantablemodel.cpp:
```diff
@@ -0,0 +1,192 @@
+        }
+    } else if (role == Qt::TextAlignmentRole) {
+        if (index.column() == Bantime)
+            return (int)(Qt::AlignRight | Qt::AlignVCenter);
```
Is int the right type here? I tried ``(QVariant)`` which also works.

Diapolo commented on src/qt/bantablemodel.cpp:
```diff
@@ -0,0 +1,192 @@
+    {
+        return createIndex(row, column, data);
+    }
+    else
```
Same as above unneeded else.

Diapolo commented on src/qt/bantablemodel.h:
```diff
@@ -0,0 +1,66 @@
+QT_END_NAMESPACE
+
+struct CCombinedBan {
+    CSubNet subnet;
```
CSubNet is in netbase.h, do we need net.h or main.h here?
Would a class CSubNet; be sufficient and include netbase.h just in the .cpp?

Diapolo commented on src/qt/bantablemodel.h:
```diff
@@ -0,0 +1,66 @@
+class QTimer;
+QT_END_NAMESPACE
+
+struct CCombinedBan {
```
We have a struct here and a std::map for similar stuff... is there a way to unify these 2?

jonasschnelli replied:
I think they should be separated and represent cores and ui form of keeping the data.

Diapolo commented on src/qt/bantablemodel.cpp:
```diff
@@ -0,0 +1,192 @@
+#include "guiconstants.h"
+#include "guiutil.h"
+
+#include "net.h"
```
This is included in the header already.

Diapolo commented on src/qt/rpcconsole.cpp:
```diff
@@ -311,23 +313,66 @@ void RPCConsole::setClientModel(ClientModel *model)
         // create context menu actions
-        QAction* disconnectAction = new QAction(tr("&Disconnect Node"), this);
+        QAction* disconnectAction   = new QAction(tr("&Disconnect Node"), this);
+        QAction* banAction1h        = new QAction(tr("&Ban Node for 1 hour"), this);
```
You are repeating some translation text, perhaps there is an efficient way to reduce that?

jonasschnelli replied:
split into two strings to reduce translation effort.

Diapolo commented on src/qt/rpcconsole.cpp:
```diff
@@ -692,6 +747,40 @@ void RPCConsole::disconnectSelectedNode()
+        SplitHostPort(nStr, port, addr);
+
+        CNode::Ban(CNetAddr(addr), bantime);
+        bannedNode->CloseSocketDisconnect();
```
Seems you reintroduced that over fDisconnect flag?

jonasschnelli replied:
This was a serious overlook. Fixed. Thanks.

Diapolo commented on src/qt/rpcconsole.cpp:
```diff
@@ -699,3 +788,10 @@ void RPCConsole::clearSelectedNode()
+
+void RPCConsole::showOrHideBanTableIfRequired()
+{
+    bool visible = clientModel->getBanTableModel()->shouldShow();
```
This could also use a NULL pointer check for clientModel IMHO.

jonasschnelli replied:
Nullpointer check added.

Diapolo commented:
Great pull IMHO, sorry for hitting you hard with that bunch of comments...

jonasschnelli commented:
Thanks @Diapolo for the review. Appricate detailed feedback.
Will fix all your points (especially the miss transaction to fDisconnect!).

jonasschnelli committed:
[Qt] bantable fix timestamp 64bit issue

jonasschnelli committed:
[Qt] bantable overhaul

- some code cleanups
- fix date formatting
- reduce header includes

jonasschnelli commented:
Prebuilt binaries: https://builds.jonasschnelli.ch/pulls/6315/

Diapolo commented on src/qt/bantablemodel.cpp:
```diff
@@ -0,0 +1,188 @@
+        }
+    } else if (role == Qt::TextAlignmentRole) {
+        if (index.column() == Bantime)
+            return (int)(Qt::AlignRight | Qt::AlignVCenter);
```
Still the question: Is ``int`` the right type here? I tried ``(QVariant)`` which also works and is more the Qt style?

jonasschnelli replied:
Right. This is also the same as we have it in PeersTableModel, it works, but i agree that it should be QVariant (as the method definition says).

Diapolo replied:
I opened https://github.com/bitcoin/bitcoin/pull/6317 and will try to keep that in sync with what we are doing here :).

Diapolo commented on src/qt/bantablemodel.h:
```diff
@@ -0,0 +1,65 @@
+#ifndef BITCOIN_QT_BANTABLEMODEL_H
+#define BITCOIN_QT_BANTABLEMODEL_H
+
+#include "net.h"
```
net.h is still there :)?

Diapolo commented on src/qt/clientmodel.cpp:
```diff
@@ -226,12 +240,19 @@ static void NotifyAlertChanged(ClientModel *clientmodel, const uint256 &hash, Ch
 
+static void BannedListChanged(ClientModel *clientmodel)
+{
+    qDebug() << "BannedListChanged";
```
You could also use something like:
``qDebug() << QString("%1: Requesting update for peer banlist").arg(__func__);`` I intend to creat a pull to make use of that format so we have the debug logging include always the correct function name :).

Diapolo commented on src/qt/rpcconsole.cpp:
```diff
@@ -8,6 +8,7 @@
 #include "clientmodel.h"
 #include "guiutil.h"
 #include "peertablemodel.h"
+#include "bantablemodel.h"
```
Not even a nit, but a please: Can you add #include "bantablemodel.h" above clientmodel :)?

Diapolo commented on src/qt/rpcconsole.cpp:
```diff
@@ -311,23 +313,66 @@ void RPCConsole::setClientModel(ClientModel *model)
+        connect(ui->peerWidget, SIGNAL(customContextMenuRequested(const QPoint&)), this, SLOT(showPeersTableContextMenu(const QPoint&)));
         connect(disconnectAction, SIGNAL(triggered()), this, SLOT(disconnectSelectedNode()));
 
+        //add a signal mapping to allow a dynamic argument
```
Nit: These comments are missing a whitespace at the front.

Diapolo commented on src/qt/rpcconsole.cpp:
```diff
@@ -335,6 +380,9 @@ void RPCConsole::setClientModel(ClientModel *model)
 
         ui->networkName->setText(QString::fromStdString(Params().NetworkIDString()));
+
+        connect(model, SIGNAL(banListChanged()), this, SLOT(showOrHideBanTableIfRequired()));
```
I'd rather group these near the banlist stuff or add some small comment, before this is getting confusing here?

Diapolo commented:
Currently integrating this into my build to test, will report back :).

Diapolo commented:
Observations:

1. Banning a node, selecting a new node (from connected peers) and then selecting a banned node leaves the details of the previous (not banned) node visible in the right area.

2. Banning a node selects the first cell in the banlist table without selecting the whole line and without a focus (cell is not marked active). I'd suggest the banned node should be selected and active entirely.

3. I'm using Tor and hidden-service addresses + netmask + banned until are that large that I'm unable to view the whole values with the default sizes.

4. The banned peers list has a heading Banned peers, perhaps the connected peers also should get a heading? I'd vote for a much smaller font size, the current one looks rather too big.

Diapolo commented:
There is also an issue with removing a peer from banned peers list when banned until is over:
![ban](https://cloud.githubusercontent.com/assets/1419649/8271710/bf357f62-1825-11e5-8739-e2b392e36dcf.png)

The peer should be removed from the list, which is not working currently.

jonasschnelli commented:
@Diapolo: for autocleaning then ban list see PR #6310

I think you point 1) is not a problem. The peer is still selected (gray) when you interact with the bantable. Point 2) and 3) are cosmetics and should not hold this PR back. Point 4) agreed.

Banning tor nodes needs testing (best first over RPC) and maybe needs more implementation. Would be good if you could see what's missing there.

jonasschnelli commented:
Tested also with Tor nodes. Works fine. @Diapolo: what issues did you encounter with tor nodes?

Diapolo commented:
@jonasschnelli The problem are incoming Tor nodes with 127.0.0.1 blocking one connected node blocks any other incoming 127.0.0.1 node. Other connected 127.0.0.1 are not affected.

Also I'd suggest splitting of the autocleaning pull from the disk-store-pull as autocleaning is uncontroversial and also needed by this one.

I also disagree on ignoring UI or UX issues for now as often such "small" or "minor" things are just forgotton or I pick them up. I'd vote for making this pull good and not just "working okay" :).

jonasschnelli commented:
Fixed @Diapolos nits.

luke-jr commented:
Can the netmask be changed to the more modern (eg) /24 /128 etc format? Also prefer to have banned peers not take up space in the peer tab... somehow.

jonasschnelli committed:
net: use CIDR notation in CSubNet::ToString()

jonasschnelli commented:
Updated.
Added a commit that uses CIDR notation (like /32)  when printing (`ToString()`) CSubNet.

UI and RPC looks now like...
![bildschirmfoto 2015-06-23 um 17 19 17](https://cloud.githubusercontent.com/assets/178464/8309724/43ddc092-19cc-11e5-8b71-db8014fd7ab3.png)
![bildschirmfoto 2015-06-23 um 17 19 21](https://cloud.githubusercontent.com/assets/178464/8309726/445898a8-19cc-11e5-9504-274cac332eb8.png)

Diapolo commented:
@jonasschnelli Thanks for your hard work on this, still there is a second round of nits open I made and also still the UX problems, the autoupdate (I have comments in the other pull also, so I'm fine if that get's some love or merged soon :)) and 127.0.0.1 handling for incoming Tor.

jonasschnelli commented:
@Diapolo: what nits are open? Font size, selecting rows, column sized are fixed.
Autocleanup is not in this PR and it will stay in #6310 because it makes most sense to clean data before writing to disk. It's not a bug if your ban list keeps a entry which has expired.
Feel free to write a patch/commit for the incoming tor node banning (separate bans by port). I won't cover it in this PR.

Diapolo commented:
@jonasschnelli Last time I saw quite some uncommented stuff in the code, even if minor. I need to resync from your repo and take another look then.

Commit:
[Qt] polish ban table

jonasschnelli commented:
Added @Diapolos polishing commit on top.

Diapolo commented:
Great, if I find the time I'll further investigate the Tor handling and also re-check the UX problems after I #6310 is merged (or after I integrated it into my build).

Diapolo commented:
I've got another pull in the pipe covering UX issues, but I need to digg in deeper...

laanwj committed:
net: Fix CIDR notation in ToString()

Only use CIDR notation if the netmask can be represented as such.

laanwj commented:
@jonasschnelli Fixed the CSubNet::ToString function for subnets that can't be represented as `/x`, please pick the top commit from https://github.com/laanwj/bitcoin/tree/2015_06_cideriad

jonasschnelli commented:
Added @laanwj commit on top to support non CIDR compatible netmasks.

Diapolo commented:
@jonasschnelli I'm asking myself if this is worth adding the core signal ``BannedListChanged`` it makes it more complex and we don't need it. We pull the list / refresh the list via timer that calls ``void BanTableModel::refresh()`` anyway. Are you okay when I rework the signal handling to be easier?

jonasschnelli commented:
@Diapolo: i extra added the signal to avoid a timer. For the peers list, a timer makes sense because we show recv/sent-bytes counters. The interval is 250ms (!). Poll updating through timers is a bad design IMO. Sometimes unavoidable.

Adding another timer (or use the same one) for the ban list would mean unnecessarily using CPU ticks. 

Well placed signals/hocks allow a software to be extendable without creating unnecessary class coupling. I think we should not be to cautious we adding new signals.

Diapolo commented:
Take a look at https://github.com/jonasschnelli/bitcoin/pull/5 IMHO this is not that bad and much easier flow-wise. We also don't have any core signals for the normal peer list.

@laanwj What do you think?

Perhaps a check could be added to verify we only pull a fresh list if the old one changed, that would reduce CPU ticks.

jonasschnelli commented:
@Diapolo: i had a look at the PR and would prefer the current solution. Your solution would make generating a relatively static ban list every 250ms which need to go over a short locks of `cs_setBanned`. 

But don't have a strong opinion on that.

Diapolo commented:
I'll try to rethink how this can be further improved, but code-wise it's much easier to understand. I respect your arguments for the core signals, but it feels like too much for this.

jonasschnelli commented:
@Diapolo: can you point out the drawbacks of adding new signal to the main signals?

Diapolo commented:
I had the impression the code-flow is overly complex, also because of updateBanlist in clientmodel, which you manually call from inside the rpcconsole. Also we have that timer code but it was disabled. The other way could be to remove the timer code and leave the core signal stuff, but also simplify clientmodel usage.

jonasschnelli committed:
[Qt] remove unused timer-code from banlistmodel.cpp

Commit:
[Qt] simplify ban list signal handling

- remove banListChanged signal from client model
- directly call clientModel->getBanTableModel()->refresh() without the way
  over clientModel->updateBanlist()

- also fix clearing peer detail window, when selecting (clicking)
  peers in the ban list

jonasschnelli commented:
Good point. Added a commit that removes the unused timer code in `bantablemodel.cpp | .h`.
The signaling approach over ClientModel is totally fine IMO. I would even say the complexity of the code is much higher with a timer than with a simple signal.

Diapolo commented:
See https://github.com/jonasschnelli/bitcoin/pull/6, which should further simplify the code a little :).

Commit:
[Qt] bantable polish

- add missing NULL pointer checks
- add better comments and reorder some code in rpcconsole.cpp
- remove unneeded leftovers in bantable.cpp
- update bantable column sizes to prevent cutting of banned until

Commit:
[Qt] add sorting for bantable

Diapolo commented:
This is progressing nicely, hope we can get some consensus on my last 3 commits in https://github.com/jonasschnelli/bitcoin/pull/6 and then get this merged :).

jonasschnelli commented:
@Diapolo: Please stop improving this PR within this PR. It's contra-productive. I can't follow all your change requests and it might crushes down this PR. Some of your things are really cool (like the sorting). But you mix everything together and that makes it hard to read and distinct between useful and not useful changes.

**Let stabilize this PR with searching and solving real problems.**
Once – and if – this is merge, you can PR your optimizations.

Diapolo commented:
Too bad... as I'm solving "real" problems ;). I won't add anything new here I promise, but it would be cool to get the things I have opened in.

jonasschnelli commented:
Added three cleanup/optimization PR from @Diapolo on top.
No it would be good to stabilize this PR.

Diapolo commented on src/qt/rpcconsole.cpp:
```diff
@@ -311,29 +312,80 @@ void RPCConsole::setClientModel(ClientModel *model)
-
-        // context menu signals
-        connect(ui->peerWidget, SIGNAL(customContextMenuRequested(const QPoint&)), this, SLOT(showMenu(const QPoint&)));
+        QAction* banAction1h      = new QAction(tr("&Ban Node for") + " " + tr("&1 hour"), this);
```
I rechecked the code and during a translation for my build I came to this.
The ``&`` is used as a shortcut when pressing ALT IMHO, so this will produce 2 shortcuts for every single menu entry. I'd suggest just removing them, because it will be hard to make them unique and still usable.

Diapolo replied:
Ping

Diapolo commented:
Would be great if you can rebase this, as the banlist.dat change was merged :). Looking forward to now get this in.

jonasschnelli committed:
[Qt] adapt QT ban option to banlist.dat changes

jonasschnelli commented:
Rebased and fix @Diapolos nits.
Also made the ban table heading text non-selectable.

jonasschnelli committed:
[Qt] reenabling hotkeys for ban context menu, use different words

- 1 (h)our
- 1 (d)ay
- 1 (w)eek
- 1 (y)ear

Commit:
[Qt] call DumpBanlist() when baning unbaning nodes

- this matches RPC call behaviour

jonasschnelli commented:
Added another commit from @Diapolo that stores the banlist.dat file after banning/unbanning over QT.

Diapolo commented:
ACK

jonasschnelli committed:
[QA] fix netbase tests because of new CSubNet::ToString() output

Diapolo commented:
Can this be rebased and get some love :)? Would be great to get this in soon...

jonasschnelli commented:
This PR doesn't need rebase and already has plenty love.
Ready for merge IMO.

jonasschnelli commented:
Here are some binaries if somebody want to test without ability to build: https://builds.jonasschnelli.ch/pulls/6315/

jonasschnelli committed:
[QA] adabt QT_NO_KEYWORDS for QT ban implementation

jonasschnelli commented:
@Diapolo: you where right, needed adaption to the post QT_NO_KEYWORDS code.
Rebased and fixed.

jgarzik commented:
ut ACK - ready to merge once conflicts fixed via rebase

jonasschnelli commented:
Rebased (trivial).

Diapolo commented:
Happy to see this moving again, I had the impression our GUI is dead...

jonasschnelli commented:
Rebased. Travis is happy now.

jtimon commented:
Concept ACK

luke-jr commented:
Why does 7f90ea78cb68c60408df85d5c653257dbc9160fe remove the DumpBanlist calls?

jonasschnelli committed:
[QT] dump banlist to disk in case of ban/unban over QT

jonasschnelli commented:
Thank @luke-jr for the precise review.
I think this sneak in over a rebase because this PR started before the banlist was persisted to disk.

Just added a commit that re-implenents the `DumpBanlist()` calls in case of banning/unbanning over QT.

luke-jr commented:
Note I did not do a precise review for this (yet?), just noticed those disappear from the previous version. ;)

Diapolo commented:
@laanwj Now if https://github.com/bitcoin/bitcoin/pull/6371 could be merged, I'd be happy!

rudirus commented:
is there some configuration file in which I can ban peers manually which haven't been connected yet?
